modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/blockstream/BlockStreamSubscription.kt

```diff
@@ -0,0 +1,137 @@
+        return when (initialized.get()) {
+            false -> {
+                initialized.set(true)
+                subscribe(
+                    context.storage.closeIf,
+                    context.storage.onFailure,
+                    context.storage.onBlock,
+                ) to getInstance(context)
+            }
+            true -> null to getInstance(context)
+        }
+    }
+
+    @JvmOverloads
+    fun <T> subscribeAndReceive(
+        closeIf: (suspend (block: VersionedBlockMessage) -> Boolean)? = null,
+        onFailure: (suspend (t: Throwable) -> Unit)? = null,
+        onBlock: (block: VersionedBlockMessage) -> Any,
+        collector: FlowCollector<T>,
+    ) = subscribe(closeIf, onFailure, onBlock).let { id ->
+        id to receive(id, collector)
+    }
+
+    fun subscribe(
+        closeIf: (suspend (block: VersionedBlockMessage) -> Boolean)? = null,
+        onFailure: (suspend (t: Throwable) -> Unit)? = null,
+        onBlock: (block: VersionedBlockMessage) -> Any,
+    ): UUID {
+        val storage = BlockStreamStorage(onBlock, closeIf, onFailure ?: context.storage.onFailure)
+        storage.channel = Channel()
+        source[storage.getId()] = storage
+        logger.debug("Block stream subscription has been expanded. Number of channels is ${source.size}")
+
+        return storage.getId()
+    }
+
+    fun <T> receive(actionId: UUID, collector: FlowCollector<T>) = launch {
+        receiveBlocking<T>(actionId).collect(collector)
+    }
+
+    fun <T> receiveBlocking(actionId: UUID): Flow<T> {
+        val storage = source[actionId] ?: throw IrohaSdkException("Flow#$actionId not found")
+        return storage.channel.cast<Channel<T>>().receiveAsFlow().catch { storage.onFailure(it) }
+    }
+
+    fun unsubscribe() {
+        stopped.set(true)
+        destroy() // singleton instance of subscription
+        logger.info("Unsubscribed from block stream")
+    }
+
+    private fun run() = launch {
+        val request = VersionedBlockSubscriptionRequest.V1(BlockSubscriptionRequest(BigInteger.valueOf(context.from)))
+
+        context.client.webSocket(
+            host = context.apiUrl.host,
+            port = context.apiUrl.port,
+            path = Iroha2Client.WS_ENDPOINT_BLOCK_STREAM,
+        ) {
+            logger.debug("WebSocket opened")
+            send(VersionedBlockSubscriptionRequest.encode(request).toFrame())
+
+            for (frame in incoming) {
+                if (stopped.get()) {
+                    this.close()
+                    source.closeAndClear()
+                    return@webSocket
+                }
+                logger.debug("Received frame: {}", frame)
+
+                val block = VersionedBlockMessage.decode(frame.readBytes())
+                source.forEach { (id, storage) ->
+                    val result = storage.onBlock(block)
+                    logger.debug("{} action result: {}", storage.getId(), result)
```

Comment:
split into 2 logs, 1 before of the action and one after

Reply:
didn't get the idea

Reply:
did you mean that's we have to log `block`?

Reply:
we might fail on `storage.onBlock(block)`, suggest just to have smth like
```
logger.debug("Executing {} action", storage.getId())
val result = storage.onBlock(block)
logger.debug("{} action result: {}", storage.getId(), result)
```